File: pyvulkan/20 - Storage Buffers/finished/frame.py
from config import *
import memory

class UBO:


    def __init__(self):

        self.view = pyrr.matrix44.create_identity(dtype=np.float32)
        self.projection = pyrr.matrix44.create_identity(dtype=np.float32)
        self.view_projection = pyrr.matrix44.create_identity(dtype=np.float32)

class SwapChainFrame:


    def __init__(self):
        
        #swapchain
        self.image = None
        self.image_view = None
        self.framebuffer = None

        self.commandbuffer = None

        #synchronization
        self.inFlight = None
        self.imageAvailable = None
        self.renderFinished = None

        #resources
        self.cameraData = UBO()
        self.uniformBuffer: memory.Buffer = None
        self.uniformBufferWriteLocation = None
        self.modelTransforms: np.ndarray = None
        self.modelBuffer: memory.Buffer = None
        self.modelBufferWriteLocation = None

        #resource descriptors
        self.uniformBufferDescriptor = None
        self.modelBufferDescriptor = None
        self.descriptorSet = None

    def make_descriptor_resources(self, logicalDevice, physicalDevice) -> None:

        #three matrices, each with 16 floats of 4 bytes each
        bufferSize = 3 * 16 * 4

        bufferInfo = memory.BufferInput()
        bufferInfo.logical_device = logicalDevice
        bufferInfo.physical_device = physicalDevice
        bufferInfo.memory_properties = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT \
            | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
        bufferInfo.size = bufferSize
        bufferInfo.usage = VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT

        self.uniformBuffer = memory.create_buffer(bufferInfo)

        self.uniformBufferWriteLocation = vkMapMemory(
            device = logicalDevice, 
            memory = self.uniformBuffer.buffer_memory, 
            offset = 0, size = bufferSize, flags = 0)
        
        """
            typedef struct VkDescriptorBufferInfo {
				VkBuffer        buffer;
				VkDeviceSize    offset;
				VkDeviceSize    range;
			} VkDescriptorBufferInfo;
        """
        self.uniformBufferDescriptor = VkDescriptorBufferInfo(
            buffer = self.uniformBuffer.buffer, offset = 0, range = bufferSize
        )

        self.modelTransforms = np.array(
            [pyrr.matrix44.create_identity() for _ in range(1024)],
            dtype = np.float32
        )

        bufferSize = 1024 * 16 * 4
        bufferInfo.size = bufferSize
        bufferInfo.usage = VK_BUFFER_USAGE_STORAGE_BUFFER_BIT

        self.modelBuffer = memory.create_buffer(bufferInfo)

        self.modelBufferWriteLocation = vkMapMemory(
            device = logicalDevice, 
            memory = self.modelBuffer.buffer_memory, 
            offset = 0, size = bufferSize, flags = 0)
        
        self.modelBufferDescriptor = VkDescriptorBufferInfo(
            buffer = self.modelBuffer.buffer, offset = 0, range = bufferSize
        )
    
    def write_descriptor_set(self, device):

        """
            typedef struct VkWriteDescriptorSet {
				VkStructureType                  sType;
				const void* pNext;
				VkDescriptorSet                  dstSet;
				uint32_t                         dstBinding;
				uint32_t                         dstArrayElement;
				uint32_t                         descriptorCount;
				VkDescriptorType                 descriptorType;
				const VkDescriptorImageInfo* pImageInfo;
				const VkDescriptorBufferInfo* pBufferInfo;
				const VkBufferView* pTexelBufferView;
			} VkWriteDescriptorSet;
        """

        descriptorWrites = [
            VkWriteDescriptorSet(
                dstSet = self.descriptorSet,
                dstBinding = 0,
                dstArrayElement = 0,
                descriptorType = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
                descriptorCount = 1,
                pBufferInfo = self.uniformBufferDescriptor
            ),
            VkWriteDescriptorSet(
                dstSet = self.descriptorSet,
                dstBinding = 1,
                dstArrayElement = 0,
                descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
                descriptorCount = 1,
                pBufferInfo = self.modelBufferDescriptor
            )
        ]
        
        vkUpdateDescriptorSets(
            device = device, 
            descriptorWriteCount = 2, 
            pDescriptorWrites = descriptorWrites, 
            descriptorCopyCount = 0, pDescriptorCopies = None
        )

File: vulkan 2022/24 - Obj Models/Finished/view/vkUtil/frame.cpp
#include "frame.h"
#include "memory.h"
#include "../vkImage/image.h"

void vkUtil::SwapChainFrame::make_descriptor_resources() {

	BufferInputChunk input;
	input.logicalDevice = logicalDevice;
	input.memoryProperties = vk::MemoryPropertyFlagBits::eHostVisible | vk::MemoryPropertyFlagBits::eHostCoherent;
	input.physicalDevice = physicalDevice;
	input.size = sizeof(UBO);
	input.usage = vk::BufferUsageFlagBits::eUniformBuffer;
	cameraDataBuffer = createBuffer(input);

	cameraDataWriteLocation = logicalDevice.mapMemory(cameraDataBuffer.bufferMemory, 0, sizeof(UBO));

	input.size = 1024 * sizeof(glm::mat4);
	input.usage = vk::BufferUsageFlagBits::eStorageBuffer;
	modelBuffer = createBuffer(input);

	modelBufferWriteLocation = logicalDevice.mapMemory(modelBuffer.bufferMemory, 0, 1024 * sizeof(glm::mat4));

	modelTransforms.reserve(1024);
	for (int i = 0; i < 1024; ++i) {
		modelTransforms.push_back(glm::mat4(1.0f));
	}

	/*
	typedef struct VkDescriptorBufferInfo {
		VkBuffer        buffer;
		VkDeviceSize    offset;
		VkDeviceSize    range;
	} VkDescriptorBufferInfo;
	*/
	uniformBufferDescriptor.buffer = cameraDataBuffer.buffer;
	uniformBufferDescriptor.offset = 0;
	uniformBufferDescriptor.range = sizeof(UBO);

	modelBufferDescriptor.buffer = modelBuffer.buffer;
	modelBufferDescriptor.offset = 0;
	modelBufferDescriptor.range = 1024 * sizeof(glm::mat4);

}

void vkUtil::SwapChainFrame::make_depth_resources() {

	depthFormat = vkImage::find_supported_format(
		physicalDevice,
		{ vk::Format::eD32Sfloat, vk::Format::eD24UnormS8Uint },
		vk::ImageTiling::eOptimal,
		vk::FormatFeatureFlagBits::eDepthStencilAttachment
	);

	vkImage::ImageInputChunk imageInfo;
	imageInfo.logicalDevice = logicalDevice;
	imageInfo.physicalDevice = physicalDevice;
	imageInfo.tiling = vk::ImageTiling::eOptimal;
	imageInfo.usage = vk::ImageUsageFlagBits::eDepthStencilAttachment;
	imageInfo.memoryProperties = vk::MemoryPropertyFlagBits::eDeviceLocal;
	imageInfo.width = width;
	imageInfo.height = height;
	imageInfo.format = depthFormat;
	depthBuffer = vkImage::make_image(imageInfo);
	depthBufferMemory = vkImage::make_image_memory(imageInfo, depthBuffer);
	depthBufferView = vkImage::make_image_view(
		logicalDevice, depthBuffer, depthFormat, vk::ImageAspectFlagBits::eDepth
	);
}

void vkUtil::SwapChainFrame::write_descriptor_set() {

	vk::WriteDescriptorSet writeInfo;
	/*
	typedef struct VkWriteDescriptorSet {
		VkStructureType                  sType;
		const void* pNext;
		VkDescriptorSet                  dstSet;
		uint32_t                         dstBinding;
		uint32_t                         dstArrayElement;
		uint32_t                         descriptorCount;
		VkDescriptorType                 descriptorType;
		const VkDescriptorImageInfo* pImageInfo;
		const VkDescriptorBufferInfo* pBufferInfo;
		const VkBufferView* pTexelBufferView;
	} VkWriteDescriptorSet;
	*/

	writeInfo.dstSet = descriptorSet;
	writeInfo.dstBinding = 0;
	writeInfo.dstArrayElement = 0; //byte offset within binding for inline uniform blocks
	writeInfo.descriptorCount = 1;
	writeInfo.descriptorType = vk::DescriptorType::eUniformBuffer;
	writeInfo.pBufferInfo = &uniformBufferDescriptor;

	logicalDevice.updateDescriptorSets(writeInfo, nullptr);

	vk::WriteDescriptorSet writeInfo2;
	writeInfo2.dstSet = descriptorSet;
	writeInfo2.dstBinding = 1;
	writeInfo2.dstArrayElement = 0; //byte offset within binding for inline uniform blocks
	writeInfo2.descriptorCount = 1;
	writeInfo2.descriptorType = vk::DescriptorType::eStorageBuffer;
	writeInfo2.pBufferInfo = &modelBufferDescriptor;

	logicalDevice.updateDescriptorSets(writeInfo2, nullptr);
}

void vkUtil::SwapChainFrame::destroy() {

	logicalDevice.destroyImageView(imageView);
	logicalDevice.destroyFramebuffer(framebuffer);
	logicalDevice.destroyFence(inFlight);
	logicalDevice.destroySemaphore(imageAvailable);
	logicalDevice.destroySemaphore(renderFinished);

	logicalDevice.unmapMemory(cameraDataBuffer.bufferMemory);
	logicalDevice.freeMemory(cameraDataBuffer.bufferMemory);
	logicalDevice.destroyBuffer(cameraDataBuffer.buffer);

	logicalDevice.unmapMemory(modelBuffer.bufferMemory);
	logicalDevice.freeMemory(modelBuffer.bufferMemory);
	logicalDevice.destroyBuffer(modelBuffer.buffer);

	logicalDevice.destroyImage(depthBuffer);
	logicalDevice.freeMemory(depthBufferMemory);
	logicalDevice.destroyImageView(depthBufferView);
}